File: sample-example.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

// Node represents a node in the tree
type Node struct {
	// The data stored at this node
	Data []byte
	// The left child of this node
	Left *Node
	// The right child of this node
	Right *Node
	// The hash of this node
	Hash []byte
}

// NewNode creates a new node with the given data
func NewNode(data []byte) *Node {
	node := &Node{
		Data: data,
		Left: nil,
		Right: nil,
	}
	node.Hash = node.calculateHash()
	return node
}

// FromSlice creates a new merkle tree from the given slice
func FromSlice(data [][]byte) *Node {
	if len(data) == 0 {
		return nil
	}
	if len(data) == 1 {
		return NewNode(data[0])
	}

	mid := len(data) / 2
	leftData := data[:mid]
	rightData := data[mid:]

	left := FromSlice(leftData)
	right := FromSlice(rightData)

	if left == nil {
		return right
	}
	if right == nil {
		return left
	}

	root := NewNode(append(left.Data, right.Data...))
	root.Left = left
	root.Right = right
	return root
}

// RootHash returns the hash of the root node
func (n *Node) RootHash() []byte {
	return n.Hash
}

// calculateHash calculates the hash of the node
func (n *Node) calculateHash() []byte {
	h := sha256.New()
	h.Write(n.Data)
	return h.Sum(nil)
}

func main() {
	data := [][]byte{
		[]byte("hello"),
		[]byte("world"),
		[]byte("this"),
		[]byte("is"),
		[]byte("a"),
		[]byte("merkle"),
		[]byte("tree"),
	}

	tree := FromSlice(data)
	rootHash := tree.RootHash()
	fmt.Printf("Root hash: %s\n", hex.EncodeToString(rootHash))
}

File: sample-example.ts
class Node<T> {
  // The data stored at this node
  data: T;
  // The left child of this node
  left: Node<T> | null;
  // The right child of this node
  right: Node<T> | null;
  // The hash of this node
  hash: number[];

  constructor(data: T) {
    this.data = data;
    this.left = null;
    this.right = null;
    this.hash = this.calculateHash(data);
  }

  // Creates a new node with the given data
  static new<T>(data: T): Node<T> {
    return new Node<T>(data);
  }

  // Recursively creates a merkle tree from the given data
  static fromArray<T>(data: T[]): Node<T> | null {
    if (data.length === 0) {
      return null;
    }
    if (data.length === 1) {
      return Node.new(data[0]);
    }

    const mid = Math.floor(data.length / 2);
    const leftData = data.slice(0, mid);
    const rightData = data.slice(mid);

    const left = Node.fromArray(leftData);
    const right = Node.fromArray(rightData);

    if (!left || !right) {
      return left || right;
    }

    const root = Node.new((left.data, right.data));
    root.left = left;
    root.right = right;
    return root;
  }

  // Returns the hash of the root node
  rootHash(): number[] {
    return this.hash;
  }

  // Calculates the hash of the given data
  private calculateHash(data: T): number[] {
    // You would need to implement a hashing function here
    // For the sake of simplicity, we will just return a random number
    return [Math.floor(Math.random() * 1000)];
  }
}

const data = [1, 2, 3, 4, 5, 6, 7, 8];
const tree = Node.fromArray(data);
const rootHash = tree!.rootHash();
console.log(`Root hash: ${rootHash}`);
